client/db/bolt/db.go

```diff
@@ -0,0 +1,535 @@
+		return bEqual(status, byteEpoch) || bEqual(status, byteBooked)
+	})
+}
+
+// AccountOrders retrieves all orders associated with the specified DEX. n = 0
+// applies no limit on number of orders returned. since = 0 is equivalent to
+// disabling the time filter, since no orders were created before before 1970.
+func (db *boltDB) AccountOrders(dex string, n int, since uint64) ([]*dexdb.MetaOrder, error) {
+	dexB := []byte(dex)
+	if n == 0 && since == 0 {
+		return db.filteredOrders(func(oBkt *bolt.Bucket) bool {
+			return bEqual(dexB, oBkt.Get(dexKey))
+		})
+	}
+	sinceB := uint64Bytes(since)
+	return db.newestOrders(n, func(oBkt *bolt.Bucket) bool {
+		timeB := oBkt.Get(updateTimeKey)
+		return bEqual(dexB, oBkt.Get(dexKey)) && bytes.Compare(timeB, sinceB) >= 0
+	})
+}
+
+// MarketOrders retrieves all orders for the specified DEX and market. n = 0
+// applies no limit on number of orders returned. since = 0 is equivalent to
+// disabling the time filter, since no orders were created before before 1970.
+func (db *boltDB) MarketOrders(dex string, base, quote uint32, n int, since uint64) ([]*dexdb.MetaOrder, error) {
+	dexB := []byte(dex)
+	baseB := uint32Bytes(base)
+	quoteB := uint32Bytes(quote)
+	if n == 0 && since == 0 {
+		return db.marketOrdersAll(dexB, baseB, quoteB)
+	}
+	return db.marketOrdersSince(dexB, baseB, quoteB, n, since)
+}
+
+// marketOrdersAll retrieves all orders for the specified DEX and market.
+func (db *boltDB) marketOrdersAll(dexB, baseB, quoteB []byte) ([]*dexdb.MetaOrder, error) {
+	return db.filteredOrders(func(oBkt *bolt.Bucket) bool {
+		return bEqual(dexB, oBkt.Get(dexKey)) && bEqual(baseB, oBkt.Get(baseKey)) &&
+			bEqual(quoteB, oBkt.Get(quoteKey))
+	})
+}
+
+// marketOrdersSince grabs market orders with optional filters for maximum count
+// and age. The sort order is always newest first. If n is 0, there is no limit
+// to the number of orders returned. If using with both n = 0, and since = 0,
+// use marketOrdersAll instead.
+func (db *boltDB) marketOrdersSince(dexB, baseB, quoteB []byte, n int, since uint64) ([]*dexdb.MetaOrder, error) {
+	sinceB := uint64Bytes(since)
+	return db.newestOrders(n, func(oBkt *bolt.Bucket) bool {
+		timeB := oBkt.Get(updateTimeKey)
+		return bEqual(dexB, oBkt.Get(dexKey)) && bEqual(baseB, oBkt.Get(baseKey)) &&
+			bEqual(quoteB, oBkt.Get(quoteKey)) && bytes.Compare(timeB, sinceB) >= 0
+	})
+}
+
+// orderTimePair is used to build an on-the-fly index to sort orders by time.
+type orderTimePair struct {
+	oid []byte
+	t   uint64
+}
+
+// newestOrders returns the n newest orders, filtered with a supplied filter
+// function. Each order's bucket is provided to the filter, and a boolean true
+// return value indicates the order should is eligible to be decoded and
+// returned.
+func (db *boltDB) newestOrders(n int, filter func(*bolt.Bucket) bool) ([]*dexdb.MetaOrder, error) {
+	var orders []*dexdb.MetaOrder
+	db.ordersView(func(master *bolt.Bucket) error {
```

Comment:
The passed function as written won't return a non-nil error, but `ordersView`/`withBucket` can if the bucket does not exists.